### DISCUSSION TITLE
How to draw some UI in Phaser3

### DISCUSSION THREAD

**Discussion by cy920820** [rexrainbow/phaser3-rex-notes#435] [General]
This may not be closely related to the current library, but as a beginner, I have indeed encountered such a problem

For example, how to locate each element and ensure that they maintain the correct display size and positioning at any resolution

**Comment by cy920820** [comment]
I want to know what the drawing approach is in this regard @rexrainbow

**Comment by rexrainbow** [comment]
It seems that size of Text and Image game objects are too small. Text and Image game object in sizer will layout by current size usually.

For text game object, might try 
- Set larger font size manually,or using [fontsize-expand-text helper method](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-text-helper/#fontsize-expand-text)

For Image game object, might try 
- Set larger scale manually. or put image into [overlapsizer](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-overview/)(see [`aspectRatio` parameter](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-overlapsizer/#add-child)), then put overlapsizer to sizer.

**Comment by cy920820** [comment]
> It seems that size of Text and Image game objects are too small. Text and Image game object in sizer will layout by current size usually.
> 
> For text game object, might try
> 
> * Set larger font size manually,or using [fontsize-expand-text helper method](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-text-helper/#fontsize-expand-text)
> 
> For Image game object, might try
> 
> * Set larger scale manually. or put image into [overlapsizer](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-overview/)(see [`aspectRatio` parameter](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-overlapsizer/#add-child)), then put overlapsizer to sizer.

Mainly, after changing the resolution, the positioning of the elements changes. Do we need to write a fixed value for the positioning coordinates?

**Reply by rexrainbow** [reply]
Assume that all children are put into nested sizers, after scale RESIZE (changing the resolution), you can [set minimum size](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-basesizer/#minimum-size) of the topmost sizer then run `topSizer.layout()` again.

**Reply by cy920820** [reply]
I have discovered a new issue, why is the text rendered by Phaser3 very blurry

**Reply by rexrainbow** [reply]
Set larger window size in game config might make text rendering more sharp.

**Reply by cy920820** [reply]
> Set larger window size in game config might make text rendering more sharp.

I have a full screen display on my phone here:
```js
    let containerWidth = container.clientWidth
    let containerHeight = container.clientHeight
    const config = {
      type: Phaser.AUTO,
      scene: [AvgBootScene, AvgGameScene],
      parent: parentID,
      resolution: window.devicePixelRatio,
      pixelArt: true
    }

    config.scale = {
      mode: Phaser.Scale.FIT,
      autoCenter: Phaser.Scale.CENTER_BOTH,
      width: containerWidth,
      height: containerHeight
    }
```

**Reply by rexrainbow** [reply]
Well, post this general issue on official discord server might get more help.

**Comment by cy920820** [comment]
> Assume that all children are put into nested sizers, after scale RESIZE (changing the resolution), you can [set minimum size](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-basesizer/#minimum-size) of the topmost sizer then run `topSizer.layout()` again.

This is the scaling configuration in the game. I have found that some other Phaser3 game source codes directly set fixed values to the width and height, which I do not have here. I set the scaled width and height based on the design draft and the width and height of the current game container, Is there any problem with me doing this? How should I set it under normal circumstances

```js
    const container = document.getElementById(parentID)
    const containerWidth = container.clientWidth
    const containerHeight = container.clientHeight
    const config = {
      type: Phaser.AUTO,
      scene: [SideScrollingScene],
      physics: {
        default: 'arcade',
        arcade: {
          gravity: { y: 0 },
          debug: false
        }
      },
      parent: parentID
    }

    let designHeight = 800
    let height = designHeight
    let width = (height * containerWidth) / containerHeight
    config.scale = {
      mode: Phaser.Scale.FIT,
      autoCenter: Phaser.Scale.CENTER_BOTH,
      width: width,
      height: height,
      parent: parentID
    }

    this.game = new Phaser.Game(config)

**Comment by rexrainbow** [comment]
With `Phaser.Scale.FIT`, the size of viewport is fixed, the resolution won't change. (Assume that you do not invoke `scene.scale.setGameSize(width, height)` during runtime).
It is a general question, post it on official discord server to get more help.

**Comment by cy920820** [comment]
@rexrainbow I would like to ask you a question, such as how to use Phaser's API to implement some rounded bubble boxes, nineSlice? Or graphics, as it involves a transition animation of height

![image](https://github.com/rexrainbow/phaser3-rex-notes/assets/21211512/9c8f5843-081b-449a-92bb-87d6f310d5d0)

**Reply by rexrainbow** [reply]
Round-rectangle shape :
- [Document](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/shape-roundrectangle/)
- [Demo](https://codepen.io/rexrainbow/pen/ZqqJjG)

Round-rectangle canvas : 
- [Document](https://rexrainbow.github.io/phaser3-rex-notes/docs/site/canvas-roundrectangle/)
- [Demo](https://codepen.io/rexrainbow/pen/rNWqaeZ)


Round-rectangle canvas can fill gradient color

**Reply by cy920820** [reply]
I found that for some conversations with variable word counts, the rounded corners have a jagged texture

**Reply by rexrainbow** [reply]
In `pixelArt: true` mode or not?

**Comment by cy920820** [comment]
When I use Nineslice, I find that the same effect cannot be achieved

This is what I have achieved, use  `this.scene.add.nineslice(0, 0, 'bgKey', 0, width, 40, 60, 60, 1, 1)`:
![image](https://github.com/rexrainbow/phaser3-rex-notes/assets/21211512/410de052-5301-4e38-be03-e4362e33edc3)

**Comment by cy920820** [comment]
use 180*150 image

![aside](https://github.com/rexrainbow/phaser3-rex-notes/assets/21211512/3f856cda-f416-475a-a1f9-e790b9e7bd8a)
